project_test/analysis_file.py

```diff
@@ -0,0 +1,207 @@
+        ConsentUtils.determine_consent_withdrawn(
+            user, data, PipelineConfiguration.SURVEY_CODING_PLANS, consent_withdrawn_key)
+        
+        # Set consent withdrawn based on stop codes from radio question answers
+        for td in data:
+            for plan in PipelineConfiguration.TEST_SHOWS_CODING_PLANS:
+                if td[f"{plan.analysis_file_key}{Codes.STOP}"] == Codes.MATRIX_1:
+                    td.append_data({consent_withdrawn_key: Codes.TRUE},
+                                    Metadata(user, Metadata.get_call_location(), time.time()))
+                
+                if plan.binary_code_scheme is not None:
+                    if td[plan.binary_coded_field]["CodeID"] == \
+                            plan.binary_code_scheme.get_code_with_control_code(Codes.STOP).code_id:
+                        td.append_data({consent_withdrawn_key: Codes.TRUE},
+                                        Metadata(user, Metadata.get_call_location(), time.time()))
+        
+        # Fold data to have one respondent per row
+        to_be_folded = []
+        for td in data:
+            to_be_folded.append(td.copy())
+
+        folded_data = FoldTracedData.fold_iterable_of_traced_data(
+            user, data, fold_id_fn=lambda td: td["uid"],
+            equal_keys=equal_keys, concat_keys=concat_keys, matrix_keys=matrix_keys, bool_keys=bool_keys,
+            binary_keys=binary_keys
+        )
+
+        # Fix-up _NA and _NC keys, which are currently being set incorrectly by
+        # FoldTracedData.fold_iterable_of_traced_data when there are multiple radio shows
+        # TODO: Update FoldTracedData to handle NA and NC correctly under multiple radio shows
+        for td in folded_data:
+            for plan in PipelineConfiguration.TEST_SHOWS_CODING_PLANS:
+                if td.get(plan.raw_field, "") != "":
+                    td.append_data({f"{plan.analysis_file_key}{Codes.TRUE_MISSING}": Codes.MATRIX_0},
+                                    Metadata(user, Metadata.get_call_location(), TimeUtils.utc_now_as_iso_string()))
+                contains_non_nc_key = False
+                for key in matrix_keys:
+                    if key.startswith(plan.analysis_file_key) and not key.endswith(Codes.NOT_CODED) \
+                        and td.get(key) == Codes.MATRIX_1:
+                        contains_non_nc_key = True
+                if not contains_non_nc_key:
+                    td.append_data({f"{plan.analysis_file_key}{Codes.NOT_CODED}": Codes.MATRIX_1},
+                                    Metadata(user, Metadata.get_call_location(), TimeUtils.utc_now_as_iso_string()))
+
+            
+            # Process consent
+            ConsentUtils.set_stopped(user, data, consent_withdrawn_key)
+            ConsentUtils.set_stopped(user, folded_data, consent_withdrawn_key)
+
+            # Output to CSV with one message per row
+            with open(csv_by_message_output_path, "w") as f:
```

Comment:
Is this at the correct level of nesting? I think you're re-writing this file many, many times

Reply:
Thank you for noticing this.